packages/node/src/login/oidc/redirectHandler/AuthCodeRedirectHandler.ts

```diff
@@ -74,8 +199,89 @@ export class AuthCodeRedirectHandler implements IRedirectHandler {
       );
     }
 
-    throw new NotImplementedError(
-      "AuthCodeRedirectHandler not implemented for Node"
+    const url = new URL(redirectUrl);
+    // The type assertion is ok, because we checked in canHandle for the presence of a state
+    const oauthState = url.searchParams.get("state") as string;
+
+    const {
+      sessionId,
+      issuerConfig,
+      codeVerifier,
+      redirectUri,
+      dpop,
+    } = await retrieveContextFromStorage(
+      oauthState,
+      this.storageUtility,
+      this.issuerConfigFetcher
+    );
+
+    const issuer = new Issuer(configToIssuerMetadata(issuerConfig));
+
+    // This should also retrieve the client from storage
+    const clientInfo: IClient = await this.clientRegistrar.getClient(
+      { sessionId },
+      issuerConfig
+    );
+
+    const client = new issuer.Client({
+      client_id: clientInfo.clientId,
+      client_secret: clientInfo.clientSecret,
+    });
+
+    const params = client.callbackParams(redirectUrl);
+    let dpopKey: JWK.ECKey;
+    let tokens: TokenSet;
+    let authFetch: typeof fetch;
+
+    if (dpop) {
+      dpopKey = await JWK.generate("EC", "P-256");
+      tokens = await client.callback(
+        redirectUri,
+        params,
+        { code_verifier: codeVerifier },
+        { DPoP: dpopKey.toJWK(true) }
+      );
+    } else {
+      tokens = await client.callback(redirectUri, params, {
+        code_verifier: codeVerifier,
+      });
+    }
+    if (tokens.access_token === undefined || tokens.id_token === undefined) {
+      // The error message is left minimal on purpose not to leak the tokens.
+      throw new Error("The IdP did not return the expected tokens.");
+    }
+    if (dpop) {
+      authFetch = await buildDpopFetch(
+        tokens.access_token,
+        tokens.refresh_token,
+        // TODO: buildDpopFetch isn't implemented yet
+        // TS thinks dpopKey isn't initialized, when it is.
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        dpopKey as any
+      );
+    } else {
+      authFetch = buildBearerFetch(tokens.access_token, tokens.refresh_token);
+    }
+
+    const webid = await deriveWebidFromTokenPayload(tokens.claims());
+
+    await saveSessionInfoInStorage(
+      this.storageUtility,
+      sessionId,
+      tokens.id_token,
+      webid,
+      "true",
+      tokens.refresh_token
     );
+
+    const sessionInfo = await this.sessionInfoManager.get(sessionId);
+    if (!sessionInfo) {
+      throw new Error(`Could not retrieve session: [${sessionId}].`);
```

Comment:
Why not? Since we don't get any information as to why (i.e. no exception, just a null/undefined response), is the only possible reason that the Session ID wasn't found, i.e.:
```suggestion
      throw new Error(`Could not find any session information associated with SessionID [${sessionId}] in storage.`);
```

Reply:
Yep, if the session cannot be retrieved at this point it means that looking up the storage for the given session ID failed.

Reply:
Resolved in 687516df762ab9f8dbbaae9605bf250ac66de6dd